core/src/main/scala/sangria/http/akka/SangriaAkkaHttp.scala

```diff
@@ -105,13 +106,71 @@ trait SangriaAkkaHttp[Input] {
       throw e
   }
 
+  /** Parse the given GraphQL query.
+    *
+    * @return [[Success]] containing the parsed query; or [[Failure]] containing a [[MalformedRequest]] exception
+    *         if no query was present, or another exception if the query failed parsing
+    */
   def prepareQuery(maybeQuery: Option[String]): Try[Document] = maybeQuery match {
     case Some(q) => QueryParser.parse(q)
-    case None => Left(MalformedRequest(MISSING_QUERY_MSG)).toTry
+    case None => Failure(MalformedRequest(MISSING_QUERY_MSG))
   }
 
-  private def extractParams: Directive[(Option[String], Option[String], Option[Input])] =
-    parameters(Symbol("query").?, Symbol("operationName").?, Symbol("variables").as[Input].?)
+  /** Extracts the standard GraphQL HTTP query parameters from an HTTP GET or POST.
+    *
+    * The extracted parameters are:
+    *   1. `query`
+    *   1. `operationName`
+    *   1. `variables`
+    *
+    * @see https://graphql.org/learn/serving-over-http/#http-methods-headers-and-body
+    */
+  private[this] val extractParams: Directive[(Option[String], Option[String], Option[Input])] =
+    parameters("query".?, "operationName".?, "variables".as[Input].?)
+
+  /** Extracts the standard GraphQL parameters from an HTTP POST.
+    *
+    * The extracted parameters are:
+    *   1. `query`
+    *   1. `operationName`
+    *   1. `variables`
+    *
+    * The parameters can come from either the HTTP query parameters or the message body,
+    * with the former taking precedence.
+    * The message body can be of either `application/json` or `application/graphql` content.
+    *
+    * This implementation differs from the
+    * [[https://graphql.org/learn/serving-over-http/#post-request informal specification]] in that it also accepts
```

Comment:
I am totally happy to match this spec and break an `application/json` POST request with `operationName` and `variables` as QS params.  Being out of spec (even informal) in this case I think is worse than flexibility. If it were up to me, I would ONLY accept `application/json` with EVERYTHING in the body 🤣

Reply:
> I am totally happy to match this spec

That would be my preference as well. I was just following the existing code.

> break an `application/json` POST request with `operationName` and variables as QS params

IDK what you mean by "break", but if I'm following the informal spec, I'd be inclined to silently ignore unrecognized URI query parameters, like these. That'd "break" _existing_ POSTs that send those params expecting them to be used. But it wouldn't "break" the request as a whole, which it would still accept, ignoring those params..

Reply:
I'm making a change to accommodate something that I just noticed in the informal spec: `query` is implied to be a non-optional field. So…if it's not optional…and "the `query` query string parameter is present"…then it still needs to be in the message body?? That can't be right. It needs to remain optional, just for this weird case where we accept it in the URI instead.